File: backend/app/schemas/vital.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional
from uuid import UUID
from datetime import datetime


class VitalCreate(BaseModel):
    """Create vital signs schema"""
    visit_id: UUID
    patient_id: UUID
    bp_systolic: Optional[int] = Field(None, ge=60, le=300)
    bp_diastolic: Optional[int] = Field(None, ge=40, le=200)
    pulse: Optional[int] = Field(None, ge=30, le=250)
    temperature: Optional[float] = Field(None, ge=35.0, le=42.0)
    respiratory_rate: Optional[int] = Field(None, ge=8, le=60)
    spo2: Optional[int] = Field(None, ge=70, le=100)
    height_cm: Optional[float] = Field(None, ge=30.0, le=250.0)
    weight_kg: Optional[float] = Field(None, ge=0.5, le=300.0)
    blood_sugar: Optional[float] = Field(None, ge=20.0, le=600.0)
    blood_sugar_type: Optional[str] = Field(None, pattern='^(fasting|random|pp)$')
    notes: Optional[str] = None
    recorded_at: Optional[datetime] = None

    @field_validator('blood_sugar_type')
    @classmethod
    def validate_blood_sugar_type(cls, v, info):
        if v and info.data.get('blood_sugar') is None:
            raise ValueError('blood_sugar_type requires blood_sugar value')
        return v


class VitalUpdate(BaseModel):
    """Update vital signs schema"""
    bp_systolic: Optional[int] = Field(None, ge=60, le=300)
    bp_diastolic: Optional[int] = Field(None, ge=40, le=200)
    pulse: Optional[int] = Field(None, ge=30, le=250)
    temperature: Optional[float] = Field(None, ge=35.0, le=42.0)
    respiratory_rate: Optional[int] = Field(None, ge=8, le=60)
    spo2: Optional[int] = Field(None, ge=70, le=100)
    height_cm: Optional[float] = Field(None, ge=30.0, le=250.0)
    weight_kg: Optional[float] = Field(None, ge=0.5, le=300.0)
    blood_sugar: Optional[float] = Field(None, ge=20.0, le=600.0)
    blood_sugar_type: Optional[str] = Field(None, pattern='^(fasting|random|pp)$')
    notes: Optional[str] = None


class VitalResponse(BaseModel):
    """Vital signs response schema"""
    id: UUID
    visit_id: UUID
    patient_id: UUID
    recorded_by: UUID
    bp_systolic: Optional[int] = None
    bp_diastolic: Optional[int] = None
    pulse: Optional[int] = None
    temperature: Optional[float] = None
    respiratory_rate: Optional[int] = None
    spo2: Optional[int] = None
    height_cm: Optional[float] = None
    weight_kg: Optional[float] = None
    bmi: Optional[float] = None
    blood_sugar: Optional[float] = None
    blood_sugar_type: Optional[str] = None
    notes: Optional[str] = None
    recorded_at: datetime
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

File: frontend/src/types/vital.ts
export interface Vital {
  id: string;
  visit_id: string;
  patient_id: string;
  recorded_by: string;
  bp_systolic?: number;
  bp_diastolic?: number;
  pulse?: number;
  temperature?: number;
  respiratory_rate?: number;
  spo2?: number;
  height_cm?: number;
  weight_kg?: number;
  bmi?: number;
  blood_sugar?: number;
  blood_sugar_type?: 'fasting' | 'random' | 'pp';
  notes?: string;
  recorded_at: string;
  created_at: string;
  updated_at: string;
}

export interface VitalCreateData {
  visit_id: string;
  patient_id: string;
  bp_systolic?: number;
  bp_diastolic?: number;
  pulse?: number;
  temperature?: number;
  respiratory_rate?: number;
  spo2?: number;
  height_cm?: number;
  weight_kg?: number;
  blood_sugar?: number;
  blood_sugar_type?: 'fasting' | 'random' | 'pp';
  notes?: string;
  recorded_at?: string;
}

export interface VitalUpdateData extends Partial<VitalCreateData> {}

// Validation ranges
export const VITAL_RANGES = {
  bp_systolic: { min: 60, max: 300, normal: [90, 120] },
  bp_diastolic: { min: 40, max: 200, normal: [60, 80] },
  pulse: { min: 30, max: 250, normal: [60, 100] },
  temperature: { min: 35.0, max: 42.0, normal: [36.5, 37.5] },
  respiratory_rate: { min: 8, max: 60, normal: [12, 20] },
  spo2: { min: 70, max: 100, normal: [95, 100] },
  height_cm: { min: 30, max: 250 },
  weight_kg: { min: 0.5, max: 300 },
  blood_sugar: { min: 20, max: 600, normal: [70, 140] },
};

export type BMICategory = 'underweight' | 'normal' | 'overweight' | 'obese';

export const getBMICategory = (bmi?: number): BMICategory | null => {
  if (!bmi) return null;
  if (bmi < 18.5) return 'underweight';
  if (bmi < 25) return 'normal';
  if (bmi < 30) return 'overweight';
  return 'obese';
};

export const getBMIColor = (category: BMICategory | null): string => {
  switch (category) {
    case 'underweight': return '#faad14';
    case 'normal': return '#52c41a';
    case 'overweight': return '#faad14';
    case 'obese': return '#ff4d4f';
    default: return '#d9d9d9';
  }
};
